authorization/token/service/token_service.go

```diff
@@ -320,6 +325,181 @@ func (s *tokenServiceImpl) Audit(ctx context.Context, identity *account.Identity
 	return &signedToken, nil
 }
 
+// Refresh checks the resource permissions in the given tokenString for the given user, and returns a
+// new RPToken (with a new expiry time and updated permissions if needed)
+func (s *tokenServiceImpl) Refresh(ctx context.Context, identity *account.Identity, tokenString string) (string, error) {
+	log.Debug(ctx, map[string]interface{}{"identity_id": identity.ID.String()}, "refreshing a user token...")
+	// Get the token manager from the context
+	manager, err := token.ReadManagerFromContext(ctx)
+	if err != nil {
+		return "", errors.NewInternalError(ctx, err)
+	}
+
+	// Now parse the token string that was passed in
+	tokenClaims, err := manager.ParseToken(ctx, tokenString)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{"error": err}, "invalid token string could not be parsed")
+		return "", errors.NewBadParameterErrorFromString("tokenString", tokenString, "invalid token string could not be parsed")
+	}
+
+	// Now that we have the identity and have parsed the token, we can see if we have a record of the token in the database
+	var tokenID uuid.UUID
+
+	// Extract the kid from the token
+	tokenID, err = uuid.FromString(tokenClaims.Id)
+	if err != nil {
+		return "", errors.NewBadParameterErrorFromString("jti", tokenClaims.Id, "invalid jti identifier - not a UUID")
+	}
+
+	loadedToken, err := s.Repositories().TokenRepository().Load(ctx, tokenID)
+	if err != nil {
+		// This is not an error per se, so we'll just log an informational message
+		log.Info(ctx, map[string]interface{}{
+			"token_id": tokenID,
+		}, "token with specified id not found")
+	}
+
+	// check if the token is still valid
+	if loadedToken != nil {
+		// Confirm that the token belongs to the current identity
+		if loadedToken.IdentityID != identity.ID {
+			return "", errors.NewUnauthorizedError("invalid token for identity")
+		}
+
+		// We now process the various token status codes in order of priority, starting with DEPROVISIONED
+		if loadedToken.HasStatus(authtoken.TOKEN_STATUS_DEPROVISIONED) {
+			return "", errors.NewUnauthorizedErrorWithCode("token deprovisioned", errors.UNAUTHORIZED_CODE_TOKEN_DEPROVISIONED)
+		}
+
+		// If the token has been revoked or the user is logged out, we respond in the same way
+		if loadedToken.HasStatus(authtoken.TOKEN_STATUS_REVOKED) || loadedToken.HasStatus(authtoken.TOKEN_STATUS_LOGGED_OUT) {
+			return "", errors.NewUnauthorizedErrorWithCode("token revoked or logged out", errors.UNAUTHORIZED_CODE_TOKEN_REVOKED)
+		}
+	}
+
+	// If we've gotten this far, it means that either no existing token was found, or the token that was found
+	// has been marked with status STALE and its privileges have changed, in either case we must generate a new token
```

Comment:
Update this comment (refresh tokens won't be stale or have privileges)

Reply:
Ah ignore this, I just realised you are only passing the old RPT token to this function, and not the refresh token.

Reply:
ok ;)